drivers/serial/uart_mcux_flexcomm.c

```diff
@@ -342,6 +379,636 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+	const struct mcux_flexcomm_config *config = dev->config;
+	struct mcux_flexcomm_data *data = dev->data;
+
+	if (config->rx_dma.dev == NULL ||
+		config->tx_dma.dev == NULL) {
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(config->rx_dma.dev) ||
+		!device_is_ready(config->tx_dma.dev)) {
+		return -ENODEV;
+	}
+
+	/* Disable DMA requests */
+	USART_EnableTxDMA(config->base, false);
+	USART_EnableRxDMA(config->base, false);
+
+	/* Route DMA requests */
+#if defined(CONFIG_SOC_SERIES_IMX_RT5XX) || defined(CONFIG_SOC_SERIES_IMX_RT6XX)
+	/* RT 3 digit uses input mux to route DMA requests from
+	 * the UART peripheral to a hardware designated DMA channel
+	 */
+	INPUTMUX_Init(INPUTMUX);
+	INPUTMUX_EnableSignal(INPUTMUX,
+		fc_uart_calc_inmux_connection(config->rx_dma.channel,
+					config->rx_dma.base), true);
+	INPUTMUX_EnableSignal(INPUTMUX,
+		fc_uart_calc_inmux_connection(config->tx_dma.channel,
+					config->tx_dma.base), true);
+	INPUTMUX_Deinit(INPUTMUX);
+#endif /* RT5xx and RT6xx */
+
+	/* Init work objects for RX and TX timeouts */
+	k_work_init_delayable(&data->tx_data.timeout_work,
+			config->tx_timeout_func);
+	k_work_init_delayable(&data->rx_data.timeout_work,
+			config->rx_timeout_func);
+
+	/* TX FIFO interrupt; configure to interrupt when empty */
+	config->base->FIFOTRIG &= ~USART_FIFOSTAT_TXLVL_MASK;
+
+	return 0;
+}
+
+#endif /* CONFIG_UART_ASYNC_API */
+
+
+static void mcux_flexcomm_isr(const struct device *dev)
+{
+	struct mcux_flexcomm_data *data = dev->data;
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	if (data->irq_callback) {
+		data->irq_callback(dev, data->irq_cb_data);
+	}
+#endif
+
+#ifdef CONFIG_UART_ASYNC_API
+	const struct mcux_flexcomm_config *config = dev->config;
+
+	/* If there is an async callback then we are using async api */
+	if (data->async_callback) {
+
+		/* Handle RX interrupt (START bit detected)
+		 * RX interrupt defeats the purpose of UART ASYNC API
+		 * because core is involved for every byte but
+		 * it is included for compatibility of applications.
+		 * There is no other way with flexcomm UART to handle
+		 * Zephyr's RX ASYNC API. However, if not using the RX
+		 * timeout (timeout is forever), then the performance is
+		 * still as might be expected.
+		 */
+		if (config->base->INTSTAT & USART_INTSTAT_START_MASK
+			&& data->async_callback) {
+
+			/* Receiving some data so reschedule timeout,
+			 * unless timeout is 0 in which case just handle
+			 * rx data now. If timeout is forever, don't do anything.
+			 */
+			if (data->rx_data.timeout == 0) {
+				flexcomm_uart_rx_update(dev);
+			} else if (data->rx_data.timeout != SYS_FOREVER_US) {
+				k_work_reschedule(&data->rx_data.timeout_work,
+						K_USEC(data->rx_data.timeout));
+			}
+
+			/* W1C start bit status bit */
+			config->base->STAT |= 0b1 & USART_STAT_START_MASK;
+		}
+
+		/* Handle TX interrupt (TXLVL = 0)
+		 * TX interrupt should happen when transfer is complete
+		 * because DMA filling TX fifo is faster than transmitter rate
```

Comment:
Just to understand this comment a bit better here- is the reason that using TXLVL=0 as the interrupt condition (rather the the TXIDLE bit) safe because the DMA engine will refill the FIFO before the USART can write enough data to empty it? I want to be sure that using TXLVL=0 as the condition does not result in the driver indicating to the user that the transfer is complete before all data was sent (especially in cases where we transmit more than 16 bytes, and the FIFO would be full and thus stall the DMA)

Reply:
yes